index.js

```diff
@@ -164,121 +166,101 @@ FeatureService.prototype.getObjectIdField = function (info) {
-                    return callback(err)
-                  }
-
-                  var minID, maxID
-                  if (idJson.error) {
-                    // DMF: if grabbing objectIDs fails fall back to guessing based on 0 and count
-                    minID = 0
-                    maxID = count
-                  } else {
-                    idJson.objectIds.sort(function (a, b) { return a - b })
-                    minID = idJson.objectIds[0]
-                    maxID = idJson.objectIds[idJson.objectIds.length - 1]
-                  }
-                  return callback(null, this._objectIdPages(minID, maxID, maxCount))
-                }.bind(this))
-
-              } catch (e) {
-                return callback(e)
-              }
-            } else {
-              var names, minId, maxId
-              var attrs = stats.features[0].attributes
-              if (stats && stats.fieldAliases) {
-                names = Object.keys(stats.fieldAliases)
-              }
-              minId = attrs.min_oid || attrs.MIN_OID || attrs[names[0]]
-              maxId = attrs.max_oid || attrs.MAX_OID || attrs[names[1]]
-
-              return callback(null, this._objectIdPages(minId, maxId, maxCount))
-            }
-          } catch (e) {
-            return callback(e)
-          }
+  this.metadata(function (err, meta) {
+    if (err) return callback(err)
+    var size = meta.size
+    var layer = meta.layer
+    this.options.objectIdField = meta.oid
+    size = Math.min(parseInt(size, 10), 1000) || 1000
+    var nPages = Math.ceil(meta.count / size)
+
+    // if the service supports paging, we can use offset to build pages
+    var canPage = layer.advancedQueryCapabilities && layer.advancedQueryCapabilities.supportsPagination
+    if (canPage) return callback(null, this._offsetPages(nPages, size))
+
+    // if the service supports statistics, we can request the maximum and minimum id to build pages
+    if (layer.supportsStatistics) {
+      this._getIdRangeFromStats(meta, function (err, stats) {
+      // if this worked then we can pagination using where clauses
+        if (!err) return callback(null, this._rangePages(stats, size))
+        // if it failed, try to request all the ids and split them into pages
+        this.layerIds(function (err, ids) {
+          // either this works or we give up
+          return callback(err, this._idPages(ids, size))
         }.bind(this))
-      } else {
-        if (count < 1000000) {
-          this.layerIds(function (err, ids) {
-            callback(err, this._idPages(ids, 250))
-          }.bind(this))
-        } else {
-          // default to sequential objectID paging starting from zero
-          return callback(null, this._objectIdPages(0, count, maxCount))
-        }
-      }
-    }.bind(this))
+      }.bind(this))
+    } else {
+      // this is the last thing we can try
+      this.layerIds(function (err, ids) {
+        callback(err, this._idPages(ids, size))
```

Comment:
Right, but if there is an error, I would guess it's not a great idea to still execute `this._idPages(ids, size)`.